File: app/src/main/java/io/ipoli/android/app/ui/UsernameValidator.java
package io.ipoli.android.app.ui;

import java.nio.charset.Charset;
import java.nio.charset.CharsetEncoder;
import java.util.regex.Pattern;

import io.ipoli.android.app.utils.StringUtils;
import io.ipoli.android.feed.persistence.FeedPersistenceService;

/**
 * Created by Venelin Valkov <venelin@curiousily.com>
 * on 7/1/17.
 */
public class UsernameValidator {

    public static final int MIN_LENGTH = 3;
    public static final int MAX_LENGTH = 20;

    public static void validate(String username, FeedPersistenceService feedPersistenceService, ResultListener resultListener) {
        if (StringUtils.isEmpty(username)) {
            resultListener.onInvalid(UsernameValidationError.EMPTY);
            return;
        }

        if(username.length() < MIN_LENGTH || username.length() > MAX_LENGTH) {
            resultListener.onInvalid(UsernameValidationError.LENGTH);
            return;
        }

        CharsetEncoder asciiEncoder = Charset.forName("US-ASCII").newEncoder();
        if (!asciiEncoder.canEncode(username)) {
            resultListener.onInvalid(UsernameValidationError.FORMAT);
            return;
        }

        Pattern p = Pattern.compile("^\\w+$");
        if (!p.matcher(username).matches()) {
            resultListener.onInvalid(UsernameValidationError.FORMAT);
            return;
        }

        feedPersistenceService.isUsernameAvailable(username, isAvailable -> {
            if (!isAvailable) {
                resultListener.onInvalid(UsernameValidationError.NOT_UNIQUE);
                return;
            }
            resultListener.onValid();
        });
    }

    public interface ResultListener {
        void onValid();

        void onInvalid(UsernameValidationError error);
    }

    public enum UsernameValidationError {
        EMPTY, NOT_UNIQUE, FORMAT, LENGTH;
    }
}

File: app/src/main/java/io/ipoli/android/player/auth/UsernameValidator.kt
package io.ipoli.android.player.auth

import io.ipoli.android.Constants
import io.ipoli.android.player.persistence.PlayerRepository
import java.nio.charset.Charset
import java.util.regex.Pattern

/**
 * Created by Polina Zhelyazkova <polina@mypoli.fun>
 * on 5/7/18.
 */
class UsernameValidator(private val playerRepository: PlayerRepository) {

    public enum class ValidationError {
        EMPTY_USERNAME,
        EXISTING_USERNAME,
        INVALID_FORMAT,
        INVALID_LENGTH
    }

    public fun validate(
        username: String
    ): ValidationError? {

        if (username.isBlank()) {
            return ValidationError.EMPTY_USERNAME
        }

        val asciiEncoder = Charset.forName("US-ASCII").newEncoder()
        if (!asciiEncoder.canEncode(username)) {
            return ValidationError.INVALID_FORMAT
        }

        val p = Pattern.compile("^\\w+$")
        if (!p.matcher(username).matches()) {
            return ValidationError.INVALID_FORMAT
        }

        if (username.length < Constants.USERNAME_MIN_LENGTH || username.length > Constants.USERNAME_MAX_LENGTH) {
            return ValidationError.INVALID_LENGTH
        }

        if (!playerRepository.isUsernameAvailable(username)) {
            return ValidationError.EXISTING_USERNAME
        }

        return null
    }
}